loader/gamedll.cpp

```diff
@@ -393,33 +646,169 @@ mm_PatchDllShutdown()
 
 	isgd_orig_shutdown = vtable_dest[isgd_shutdown_index];
 	vtable_dest[isgd_shutdown_index] = vtable_src[mfp.vtblindex];
 }
 
+#if defined _WIN32
+static void
+mm_PatchAllowDedicated(bool patch)
+{
+	void **vtable_src;
+	void **vtable_dest;
+	SourceHook::MemFuncInfo mfp;
+
+	SourceHook::GetFuncInfo(&ISource2ServerConfig::AllowDedicatedServers, mfp);
+
+	assert(mfp.isVirtual);
+	assert(mfp.thisptroffs == 0);
+	assert(mfp.vtbloffs == 0);
+
+	vtable_src = (void **) *(void **) &is2sc_thunk;
+	vtable_dest = (void **) *(void **) config_iface;
+
+	SourceHook::SetMemAccess(&vtable_dest[is2sc_allowdedi_index],
+		sizeof(void*),
+		SH_MEM_READ | SH_MEM_WRITE | SH_MEM_EXEC);
+
+	if (patch)
+	{
+		assert(is2sc_orig_allowdedi == NULL);
+		is2sc_orig_allowdedi = vtable_dest[is2sc_allowdedi_index];
+		vtable_dest[is2sc_allowdedi_index] = vtable_src[mfp.vtblindex];
+	}
+	else
+	{
+		assert(is2sc_orig_allowdedi != NULL);
+		vtable_dest[is2sc_allowdedi_index] = is2sc_orig_allowdedi;
+		is2sc_orig_allowdedi = NULL;
+	}
+}
+#endif
+
 static void
-mm_PrepForGameLoad()
+mm_PatchConnect(bool patch)
 {
-	mm_PatchDllInit(true);
+	void **vtable_src;
+	void **vtable_dest;
+	SourceHook::MemFuncInfo mfp;
+
+	SourceHook::GetFuncInfo(&ISource2ServerConfig::Connect, mfp);
+
+	assert(mfp.isVirtual);
+	assert(mfp.thisptroffs == 0);
+	assert(mfp.vtbloffs == 0);
+
+	vtable_src = (void **) *(void **) &is2sc_thunk;
+	vtable_dest = (void **) *(void **) config_iface;
+
+	SourceHook::SetMemAccess(&vtable_dest[mfp.vtblindex],
+		sizeof(void*),
+		SH_MEM_READ | SH_MEM_WRITE | SH_MEM_EXEC);
+
+	if (patch)
+	{
+		assert(is2sc_orig_connect == NULL);
+		is2sc_orig_connect = vtable_dest[mfp.vtblindex];
+		vtable_dest[mfp.vtblindex] = vtable_src[mfp.vtblindex];
+	}
+	else
+	{
+		assert(is2sc_orig_connect != NULL);
+		vtable_dest[mfp.vtblindex] = is2sc_orig_connect;
+		is2sc_orig_connect = NULL;
+	}
 }
 
+
 void *
 mm_GameDllRequest(const char *name, int *ret)
 {
-	if (gamedll_lib != NULL && gamedll_bridge == NULL)
+	if (strncmp(name, "Source2ServerConfig", 19) == 0)
 	{
-		return gamedll_qvi(name, ret);
+		g_is_source2 = true;
+		if (!mm_DetectGameInformation())
+		{
+			if (ret != NULL)
+				*ret = 1;
+			return NULL;
+		}
+
+		void *lib;
+		char error[255];
+		void *ptr = NULL;
+		QueryValveInterface qvi;
+		for (unsigned int i = 0; i < gamedll_path_count; i++)
+		{
+			if (gamedll_libs[i] == NULL)
+			{
+				lib = mm_LoadLibrary(gamedll_paths[i], error, sizeof(error));
+				if (lib == NULL)
+					continue;
+				gamedll_libs[i] = lib;
+			}
+			lib = gamedll_libs[i];
+			qvi = (QueryValveInterface)mm_GetLibAddress(lib, "CreateInterface");
+			if (qvi == NULL)
+				continue;
+			ptr = qvi(name, ret);
+			if (ptr != NULL)
+			{
+				gamedll_libs[i] = NULL;
+				break;
+			}
+		}
+
+		if (ptr != NULL)
+		{
+			mm_FreeCachedLibraries();
+			gamedll_lib = lib;
+			config_iface = (ISource2ServerConfig *) ptr;
+			gamedll_qvi = qvi;
+
+			mm_PatchConnect(true);
```

Comment:
I'm not saying this is wrong, this is pretty required; but we're making a "new" delineation here that MM:S is actively patching the game, enabling clients to connect and the server to even start as a dedicated server. More of a "okay this is changing with MM:S" instead of a "hey whoa there" as things like this have traditionally lived in a plugin to the best of my knowledge.